File: apogee-backend/apogee/services/prompt_service.py
from functools import lru_cache
from pathlib import Path
from string import Template

SUMMARY_STYLES = {
    "bullets": (
        "Return only the final answer.\n\n"
        "Rules:\n"
        "- Output 8-14 concise bullet points.\n"
        "- Each bullet must be on its own line.\n"
        "- Do not write any introduction.\n"
        "- Do not write any heading.\n"
        "- Do not write any conclusion.\n"
        "- Do not explain what you are doing.\n"
        "- Do not prefix the output with phrases like "
        "'Here is the summary', 'Summary:', or similar.\n"
        "- Output only the bullet points."
    ),

    "sentences": (
        "Return only the final answer.\n\n"
        "Rules:\n"
        "- Output exactly 7-10 concise sentences.\n"
        "- Put each sentence on a separate line.\n"
        "- Do not use bullets.\n"
        "- Do not use numbering.\n"
        "- Do not write a paragraph.\n"
        "- Do not write any introduction.\n"
        "- Do not write any heading.\n"
        "- Do not write any conclusion.\n"
        "- Do not prefix the response with phrases like "
        "'Here is the summary', 'Summary:', "
        "'Below is a summary', or similar.\n"
        "- Output only the sentences."
    ),

    "paragraphs": (
        "Return only the final answer.\n\n"
        "Rules:\n"
        "- Output one concise paragraph containing 10-15 sentences.\n"
        "- Do not use bullets.\n"
        "- Do not use numbering.\n"
        "- Do not add a heading.\n"
        "- Do not write an introduction.\n"
        "- Do not write a conclusion.\n"
        "- Do not prefix the response with phrases like "
        "'Here is the summary', 'Summary:', or similar.\n"
        "- Output only the paragraph."
    ),
}

PROMPTS_DIR = (
    Path(__file__).resolve().parent.parent
    / "prompts"
)


@lru_cache(maxsize=None)
def load_prompt(name):
    """Load a prompt template from disk (cached after first read)."""
    path = PROMPTS_DIR / f"{name}.txt"
    return path.read_text(encoding="utf-8")

def build_summary_prompt(title, url, content, mode):
    template_str = load_prompt("summarize")
    style = SUMMARY_STYLES.get(mode, SUMMARY_STYLES["bullets"])

    return Template(template_str).safe_substitute(
        title=title,
        url=url,
        content=content,
        style=style,
    )

def build_answer_prompt(title, url, content, question):
    template_str = load_prompt("answer")
    return Template(template_str).safe_substitute(
        title=title,
        url=url,
        content=content,
        question=question,
    )


def build_suggested_questions_prompt(title, url, summary):
    template_str = load_prompt("suggest_questions")
    return Template(template_str).safe_substitute(
        title=title,
        url=url,
        summary=summary,
    )

File: apogee-backend/src/services/promptService.js
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROMPTS_DIR = path.join(__dirname, "..", "prompts");

const promptCache = new Map();

/** Load a prompt template from disk (cached after first read). */
function loadPrompt(name) {
  if (!promptCache.has(name)) {
    const filePath = path.join(PROMPTS_DIR, `${name}.txt`);
    promptCache.set(name, readFileSync(filePath, "utf-8"));
  }
  return promptCache.get(name);
}

// Mirrors Python's string.Template.safe_substitute: replaces $identifier /
// ${identifier} with values from `substitutions` when present, leaves the
// placeholder untouched otherwise, and treats "$$" as a literal "$".
const PLACEHOLDER = /\$(\$|[_a-zA-Z][_a-zA-Z0-9]*|\{[_a-zA-Z][_a-zA-Z0-9]*\}|)/g;

function safeSubstitute(template, substitutions) {
  return template.replace(PLACEHOLDER, (match, token) => {
    if (token === "$") return "$";
    if (token === "") return match;
    const key = token.startsWith("{") ? token.slice(1, -1) : token;
    return Object.prototype.hasOwnProperty.call(substitutions, key)
      ? String(substitutions[key])
      : match;
  });
}

const SUMMARY_STYLES = {
  bullets: [
    "Return only the final answer.",
    "",
    "Rules:",
    "- Output 8-14 concise bullet points.",
    "- Each bullet must be on its own line.",
    "- Do not write any introduction.",
    "- Do not write any heading.",
    "- Do not write any conclusion.",
    "- Do not explain what you are doing.",
    "- Do not prefix the output with phrases like " +
      "'Here is the summary', 'Summary:', or similar.",
    "- Output only the bullet points.",
  ].join("\n"),

  sentences: [
    "Return only the final answer.",
    "",
    "Rules:",
    "- Output exactly 7-10 concise sentences.",
    "- Put each sentence on a separate line.",
    "- Do not use bullets.",
    "- Do not use numbering.",
    "- Do not write a paragraph.",
    "- Do not write any introduction.",
    "- Do not write any heading.",
    "- Do not write any conclusion.",
    "- Do not prefix the response with phrases like " +
      "'Here is the summary', 'Summary:', " +
      "'Below is a summary', or similar.",
    "- Output only the sentences.",
  ].join("\n"),

  paragraphs: [
    "Return only the final answer.",
    "",
    "Rules:",
    "- Output one concise paragraph containing 10-15 sentences.",
    "- Do not use bullets.",
    "- Do not use numbering.",
    "- Do not add a heading.",
    "- Do not write an introduction.",
    "- Do not write a conclusion.",
    "- Do not prefix the response with phrases like " +
      "'Here is the summary', 'Summary:', or similar.",
    "- Output only the paragraph.",
  ].join("\n"),
};

export function buildSummaryPrompt({ title, url, content, mode }) {
  const template = loadPrompt("summarize");
  const style = SUMMARY_STYLES[mode] ?? SUMMARY_STYLES.bullets;
  return safeSubstitute(template, { title, url, content, style });
}

export function buildAnswerPrompt({ title, url, content, question }) {
  const template = loadPrompt("answer");
  return safeSubstitute(template, { title, url, content, question });
}

export function buildSuggestedQuestionsPrompt({ title, url, summary }) {
  const template = loadPrompt("suggest_questions");
  return safeSubstitute(template, { title, url, summary });
}
